### ISSUE TITLE
`i.mjs` with only `i = 0;` isn't throwing an error

### ISSUE THREAD

**Issue by kungfooman** [microsoft/TypeScript#54497]
# Bug Report

![image](https://github.com/microsoft/TypeScript/assets/5236548/29a590e6-f7a2-492f-ba06-6cd80fe59935)

### 🔎 Search Terms

Accessing undefined variable in ModuleJS/mjs file is not causing any error

### 💻 Code

**i.mjs**

```js
i = 0;
```

someOtherRandomFileInRepo.js
```
var i = 0;
```

### 🙁 Actual behavior

Doesn't throw an error.

### 🙂 Expected behavior

Throw an error.

I'm using *.mjs exactly because I want some strict behaviour, but why isn't this erroring? Is there any way to enable it via `jsconfig.json` or is this a TS bug?

I always try to already make my `jsconfig.json` as strict as possible:

```json
{
  "compilerOptions": {
    "target" : "es2022",
    "module" : "es2022",
    "strict" : true,
    "strictNullChecks": true,
    "checkJs": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noFallthroughCasesInSwitch": true,
    "noPropertyAccessFromIndexSignature": true,
    "forceConsistentCasingInFileNames": true,
    "moduleResolution": "node",
  },
  "include": [
    "./*",
    "./src/*",
    "./src/**/*",
    "./src/**/**/*",
    "./src/**/**/**/*",
  ]
}
```

**Comment by MartinJohns** [comment]
I'm not completely up to date with the mjs files, but I believe due to the lack of any export/import statement in your `someOtherRandomFileInRepo.js` file it's considered a global file, and using `var` there will pollute the global namespace. Your `i.mjs` has access to this global namespace.

**Comment by kungfooman** [comment]
@MartinJohns So some random file is able to disable parts of type safety? I want my `*.mjs` to access only and exactly what I declare on top of it, so I still believe this is a bug, or should this be considered "good and standard behaviour"?

**Comment by fatcerberus** [comment]
> I want my `*.mjs` to access only and exactly what I declare on top of it

Interpreted strictly, this would mean you wouldn't be able to use e.g. APIs defined in `lib.dom.d.ts` in an ES module.  TS thinks the `.js` file will run in the global scope at runtime, in which case the variable(s) it declares would be accessible to *all* files at runtime, even modules.  So yes, that *is* expected behavior.  However, since it's unlikely the `.js` file actually *will* be run as global code, you probably want to put `moduleDetection: "force"` in your tsconfig.

**Comment by kungfooman** [comment]
> However, since it's unlikely the `.js` file actually _will_ be run as global code, you probably want to put `moduleDetection: "force"` in your tsconfig.

Thank you very much, always nice to read your insights! :1st_place_medal: 

Yes, I see your point about defined APIs, isn't that also the point of https://www.typescriptlang.org/tsconfig#lib?

For example: `"lib": ["ESNext", "DOM", "DOM.Iterable"]`

Anyway, your suggestion works! The only problem left is that this doesn't work in `jsconfig.json`. Why is this only possible in `tsconfig.json`?

Kinda feels like that `moduleDetection: "force"` should be the default behaviour for `*.mjs` files aswell :thinking: I noticed this error only as "runtime bug".

**Comment by RyanCavanaugh** [comment]
The setting `moduleDetection: force` has no effect on the contents of an MJS file, because a file with that extension is already a module. I think what you're trying to say is that any compilation containing an MJS file at all would have that setting set, but I think that would be pretty disruptive and confusing.

Are you sure that this doesn't work in jsconfig.json? I'm not convinced that we have some weird setting-specific bug here

**Comment by kungfooman** [comment]
@RyanCavanaugh Yes, I tried it and there was no effect, auto-suggestion also fails:

![image](https://github.com/microsoft/TypeScript/assets/5236548/50e516e6-19f6-4154-ab65-5a986cc9f0ef)

It only works inside `tsconfig.json`, but I'm using pure MJS, so I don't know why I should have a `tsconfig.json`.

**Comment by RyanCavanaugh** [comment]
Works for me (error appears iff I set moduleDetection: force, as expected).

![image](https://github.com/microsoft/TypeScript/assets/6685088/876ce018-2c31-4cb9-9958-354372e2f743)

**Comment by kungfooman** [comment]
@RyanCavanaugh Thank you for testing, I tried updating VSCode but it still fails. I also investigated the schemas a bit:

https://json.schemastore.org/tsconfig
https://json.schemastore.org/jsconfig

The `jsconfig` one doesn't contain `moduleDetection`:

https://github.com/SchemaStore/schemastore/blob/cae837753663d9338e01a7aedc098babbc837af6/src/schemas/json/tsconfig.json#L910-L920

https://github.com/SchemaStore/schemastore/blob/cae837753663d9338e01a7aedc098babbc837af6/src/schemas/json/jsconfig.json#L898-L904

So I'm not sure what's going on, maybe the stable version of TypeScript doesn't support this yet and you are running the latest version? Or we are using different schemas. :thinking:

**Comment by fatcerberus** [comment]
The schema is unrelated - it just tells VSCode which options are valid in the file so it can offer suggestions and show errors for invalid settings.  That's a separate concern to whether *TypeScript*--and by extension the TypeScript language service--supports the option.

**Comment by kungfooman** [comment]
> That's a separate concern to whether _TypeScript_--and by extension the TypeScript language service--supports the option.

Right, regardless of schema offering the suggestion, `moduleDetection` didn't work in `jsconfig.json`.

I just realized why this happened, the sheer existence of `tsconfig.json` made the option in `jsconfig.json` useless:

![image](https://github.com/microsoft/TypeScript/assets/5236548/53395870-fa8c-4f99-959c-f52a6265bbde)

I don't know why, but removing/renaming `tsconfig.json` fixes the issue. :+1: 

Thank you for all the help, this option finally works in `jsconfig.json` and TS will catch these errors now :zany_face:

**Comment by fatcerberus** [comment]
Yeah, `tsconfig.json` takes precedence over `jsconfig.json`, so that’s expected

**Comment by kungfooman** [comment]
> Yeah, `tsconfig.json` takes precedence over `jsconfig.json`, so that’s expected

Only strange thing is that there is "nothing" to take precedence over, right?

**Comment by fatcerberus** [comment]
Well, the files don’t have their options merged, so if the file that has precedence is empty, you just get defaults.
